### PR TITLE
Support sharding via envoy

### PR DESCRIPTION
**Pull request by szabado**

### CODE REVIEW

File: buckets/redis/bucket_factory.go [hunk 1]
```diff
@@ -92,29 +143,29 @@ func (bf *bucketFactory) Init(cfg *pbconfig.ServiceConfig) {
...
 }
 
 func (bf *bucketFactory) connectToRedisLocked() {
 	// Set up connection to Redis
 	bf.client = redis.NewClient(bf.redisOpts)
 
-	t, err := bf.client.Time().Result()
+	_, err := bf.client.Touch("areYouAlive?").Result()
 	if err != nil {
 		logging.Printf("Cannot connect to Redis. TIME returned %v", err)
```
Comment (spenceral) [inline buckets/redis/bucket_factory.go]: update message

File: buckets/redis/bucket_factory.go [hunk 2]
```diff
@@ -92,29 +143,29 @@ func (bf *bucketFactory) Init(cfg *pbconfig.ServiceConfig) {
...
-	bf.scriptSHA = loadScript(bf.client)
 }
 
 func (bf *bucketFactory) reconnectToRedis(oldClient *redis.Client) {
 	bf.Lock()
 	defer bf.Unlock()
 
 	// Always close connections on errors to prevent results leaking.
-	if err := bf.client.Close(); !isRedisClientClosedError(err) {
+	if err := bf.client.Close(); err != nil && !isRedisClientClosedError(err) {
```
Comment (spenceral) [inline buckets/redis/bucket_factory.go]: Why did you add this nil check?

File: buckets/redis/bucket.go [hunk 3]
```diff
@@ -80,9 +80,9 @@ func (a *abstractBucket) Take(ctx context.Context, requested int64, maxWaitTime
 }
 
 func (a *abstractBucket) takeFromRedis(ctx context.Context, client *redis.Client, args []interface{}) *redis.Cmd {
-	span, ctx := opentracing.StartSpanFromContext(ctx, "EvalSha")
+	span, ctx := opentracing.StartSpanFromContext(ctx, "script.Run")
```
Comment (spenceral) [inline buckets/redis/bucket.go]: Have you done any benchmarking with script.Run compared to executing a pre-loaded script?

Comment (szabado) [reply buckets/redis/bucket.go]: I haven't. `script.Run` should have the same amortized performance as `EvalSha`, or at least comparable performance.

Redis will [cache scripts](https://redis.io/commands/eval#script-cache-semantics) that are run with `EVAL`, so initial requests when redis comes up will fail, but then all subsequent `EVALSHA`'s will succeed until redis restarts/the script changes. It's a small penalty for the requests after a server restarts, but I think it should be _relatively_ minimal.

Comment (szabado) [reply buckets/redis/bucket.go]: `script.Run` calls `EVALSHA`, followed by `EVAL` if it fails ([implementation](https://github.com/go-redis/redis/blob/master/script.go#L56-L62)).

File: buckets/redis/bucket_factory.go (hunk 2 above)
Comment (szabado) [reply buckets/redis/bucket_factory.go]: `isRedisClientClosedError` can be inlined to `err != nil && err.Error() == "bla"`. Simplifying this with some [boolean algebra](https://www.electronics-tutorials.ws/boolean/bool_6.html):
`err != nil && !(err != nil && err.Error == "bla")` -> de Morgan's law
`err != nil && (err == nil || err.Error != "bla")` -> distributive law
`(err != nil && err == nil) || (err != nil && err.Error != "bla")` -> complement law
`err != nil && err.Error != "bla"`

versus

`!(err != nil && err.Error == "bla")` -> de Morgan's law
`err == nil || err.Error != "bla"`

Basically without this check, it doesn't actually do what it looks like it does.

File: buckets/redis/bucket_factory.go (hunk 1 above)
Comment (szabado) [reply buckets/redis/bucket_factory.go]: done

File: buckets/redis/bucket_factory.go (hunk 2 above)
Comment (szabado) [reply buckets/redis/bucket_factory.go]: I introduced this bug in an earlier commit, trying to fix a poorly named function

Comment (spenceral) [reply buckets/redis/bucket_factory.go]: I missed the `!` before isRedisClientClosed.. ignore me.

File: buckets/redis/bucket.go (hunk 3 above)
Comment (spenceral) [reply buckets/redis/bucket.go]: Nice
